src/main/java/com/epam/esm/service/CertificateService.java

```diff
@@ -0,0 +1,226 @@
+        }
+    }
+
+    //TODO apply transactional
+    @Transactional
+    public void deleteGiftCertificate(Long certificateId) {
+
+        if(!certificateRepository.existsById(certificateId)){
+            throw new CustomizedException(CERTIFICATE_WITH_ID_NOT_FOUND.formatted(certificateId), ErrorCode.CERTIFICATE_NOT_FOUND);
+        }
+        try {
+            certificateRepository.deleteById(certificateId);
+        } catch (DataAccessException ex) {
+            throw new CustomizedException("Database error during deleting certificate with id " + certificateId, ErrorCode.CERTIFICATE_DATABASE_ERROR, ex);
+        }
+    }
+    @Transactional
+    public CertificateResponseDTO updateGiftCertificate(@NonNull Long certificateId, GiftCertificate updates, List<Long> newTagIdsList) {
+
+        try {
+            Optional<GiftCertificate> optCertificate = certificateRepository.findById(
+                certificateId);
+
+            if (optCertificate.isEmpty()) {
+                throw new CustomizedException(
+                    "Gift certificate not found with id: " + certificateId,
+                    ErrorCode.CERTIFICATE_NOT_FOUND);
+            }
+
+            GiftCertificate existingCertificate = optCertificate.get();
+
+            if (updates.getName() != null) {
+                existingCertificate.setName(updates.getName());
+            }
+            if (updates.getDescription() != null) {
+                existingCertificate.setDescription(updates.getDescription());
+            }
+            if (updates.getPrice() != null) {
+                existingCertificate.setPrice(updates.getPrice());
+            }
+            if (updates.getDuration() != null) {
+                existingCertificate.setDuration(updates.getDuration());
+            }
+            if (newTagIdsList != null) {
+                List<Tag> uniqueTags = tagRepository.findAllByIdIn(newTagIdsList);
+                existingCertificate.setTags(uniqueTags);
+            }
+
+            GiftCertificate updatedCertificate = certificateRepository.save(existingCertificate);
+
+            return convertToCertificateDTO(updatedCertificate);
+        } catch (DataAccessException ex){
+            throw new CustomizedException("Database error during update certificate with id " + certificateId, ErrorCode.CERTIFICATE_DATABASE_ERROR, ex);
+
+        }
+    }
+
+    private Optional<ErrorDTO> validateCertificateRequest(GiftCertificate giftCertificate, List<Long> tagIds) {
+
+        Optional<String> validationMessage = validateRequest(giftCertificate);
+        if (validationMessage.isPresent()) {
+            ErrorDTO errorResponse = new ErrorDTO(validationMessage.get(), CERTIFICATE_BAD_REQUEST.getErrorCode());
+            return Optional.of(errorResponse);
+        }
+
+        List<Tag> existingTags = tagRepository.findAllById(tagIds);
```

Comment:
Well, actually I forgot to replace findAllByIdIn, because from what I understand, using findAllById is enough and don't need to create an additional method in the repository.